Lib/test/test_capi/test_codecs.py

```diff
@@ -780,33 +842,70 @@ def test_codec_lookup_error(self):
-            UnicodeEncodeError('bad', 'xyz123', 1, 4, 'reason'),
-        ]
-
-        strict_handler = _testcapi.codec_strict_errors
+    def test_codec_strict_errors_handler(self):
+        handler = _testcapi.codec_strict_errors
+        for exc in self.all_unicode_errors + self.bad_unicode_errors:
+            with self.subTest(handler=handler, exc=exc):
+                self.assertRaises(type(exc), handler, exc)
+
+    def test_codec_ignore_errors_handler(self):
+        handler = _testcapi.codec_ignore_errors
+        all_exceptions = self.all_unicode_errors
+        bad_exceptions = self.bad_unicode_errors
+        self.do_test_codec_errors_handler(handler, all_exceptions, bad_exceptions)
+
+    def test_codec_replace_errors_handler(self):
+        handler = _testcapi.codec_replace_errors
+        all_exceptions = self.all_unicode_errors
+        bad_exceptions = self.bad_unicode_errors
+        self.do_test_codec_errors_handler(handler, all_exceptions, bad_exceptions)
+
+    def test_codec_xmlcharrefreplace_errors_handler(self):
+        handler = _testcapi.codec_xmlcharrefreplace_errors
+        exceptions = self.unicode_encode_errors
+        bad_exceptions = (
+            self.bad_unicode_errors
+            + tuple(e for e in self.all_unicode_errors if e not in exceptions)
+        )
+        self.do_test_codec_errors_handler(handler, exceptions, bad_exceptions)
+
+    def test_codec_backslashreplace_errors_handler(self):
+        handler = _testcapi.codec_backslashreplace_errors
+        exceptions = self.all_unicode_errors
+        bad_exceptions = self.bad_unicode_errors
+        self.do_test_codec_errors_handler(handler, exceptions, bad_exceptions)
+
+    def test_codec_namereplace_errors_handler(self):
+        handler = _testlimitedcapi.codec_namereplace_errors
+        exceptions = self.unicode_encode_errors
+        bad_exceptions = (
+            self.bad_unicode_errors
+            + tuple(e for e in self.all_unicode_errors if e not in exceptions)
+        )
+        self.do_test_codec_errors_handler(handler, exceptions, bad_exceptions)
+
+    def do_test_codec_errors_handler(self, handler, exceptions, bad_exceptions):
+        at_least_one = False
         for exc in exceptions:
-            with self.subTest(handler=strict_handler, exc=exc):
-                self.assertRaises(UnicodeEncodeError, strict_handler, exc)
-
-        for handler in [
-            _testcapi.codec_ignore_errors,
-            _testcapi.codec_replace_errors,
-            _testcapi.codec_xmlcharrefreplace_errors,
-            _testlimitedcapi.codec_namereplace_errors,
-        ]:
-            for exc in exceptions:
-                with self.subTest(handler=handler, exc=exc):
-                    self.assertIsInstance(handler(exc), tuple)
+            # See https://github.com/python/cpython/issues/123378 and related
+            # discussion and issues for details.
+            if exc.may_crash:
+                continue
```

Comment:
Rather than subclassing, consider putting the may_crash logic in a method like:
```
    def _may_crash(self, exc):
        # See https://github.com/python/cpython/issues/123378 and related
        # discussion and issues for details.
        start = exc.start
        end = exc.end
        message = exc.object
        match exc:
            case UnicodeEncodeError():
                return end < start or (end - start) >= len(message)
            case UnicodeDecodeError():
                return end < start
            case UnicodeTranslateError():
                return end <= start or (end - start) >= len(message)
        return False
```
That'll be easier to remove when the time comes.

Reply:
Unfortunately, using `exc.start` uses the "bad" getter so there are some corner cases that would be incorrectly detected due to the clipping. Once we fixed the getter, we can use that approach.

Reply:
Hmm, which cases are that?

Reply:
IIRC, when start > length and or when start < 0 (I really want to store start as a negative value and the constructor does not clip values at all) or when end < 0 and end >= length.

Reply:
Nevermind, actually the getters are not called! the members are directly accessed on the structure. However I found an issue with this:

```py
>>> x = UnicodeDecodeError('utf-8', b'00000', -9, 2, 'reason')
>>> x
UnicodeDecodeError('utf-8', b'00000', -9, 2, 'reason')
>>> x.start = -3
>>> x
UnicodeDecodeError('utf-8', b'00000', -9, 2, 'reason')
```

There is no change at all in the representation! Is it a bug?

Reply:
It's fine, the message doesn't need to be dynamic.